File: fdr_exercises/cpp/custom_joint_publisher.cpp
#include <custom_joint_publisher.h>
#include <math.h>

void send_des_jstate(const JointStateVector & joint_pos)
{
    std::cout << "q_des " << joint_pos.transpose() << std::endl;
    if (control_type=="position")
    {
        for (int i = 0; i < joint_pos.size(); i++)
        {
          jointState_msg_position.data[i] = joint_pos[i];
        }

        pub_des_jstate.publish(jointState_msg_position);


    } else {
        for (int i = 0; i < joint_pos.size(); i++)
        {
          jointState_msg_torque.position[i] = joint_pos[i];
          jointState_msg_torque.velocity[i] = 0.0;
          jointState_msg_torque.effort[i] = 0.0;
        }

        pub_des_jstate.publish(jointState_msg_torque);
    }

/*   if (pub_des_jstate_sim_rt->trylock())
  {
    pub_des_jstate_sim_rt->msg_ = jointState_msg;
    pub_des_jstate_sim_rt->unlockAndPublish();
  } */
}


void initFilter(const JointStateVector & joint_pos)
{
        filter_1 = joint_pos;
        filter_2 = joint_pos;
}

JointStateVector secondOrderFilter(const JointStateVector & input, const double rate, const double settling_time)
{

        double dt = 1 / rate;
        double gain =  dt / (0.1*settling_time + dt);
        filter_1 = (1 - gain) * filter_1 + gain * input;
        filter_2 = (1 - gain) * filter_2 + gain *filter_1;
        return filter_2;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "custom_joint_publisher");
  ros::NodeHandle node;

  //pub_des_jstate_sim_rt.reset(new realtime_tools::RealtimePublisher<sensor_msgs::JointState>(node, "/command", 1));
  node.getParam("/control_type", control_type);



  if (control_type=="position")
  {
      pub_des_jstate = node.advertise<std_msgs::Float64MultiArray>("/ur5/joint_group_pos_controller/command", 1);

  } else {
      pub_des_jstate = node.advertise<sensor_msgs::JointState>("/command", 1);
  }

  ros::Rate loop_rate(loop_frequency);

  jointState_msg_torque.position.resize(6);
  jointState_msg_torque.velocity.resize(6);
  jointState_msg_torque.effort.resize(6);
  jointState_msg_position.data.resize(6);

  q_des0 << -0.3223527113543909, -0.7805794638446351, -2.5675506591796875, -1.6347843609251917, -1.5715253988849085, -1.0017417112933558;
  initFilter(q_des0);

  JointStateVector amp;
  JointStateVector freq;
  amp << 0.3, 0.0, 0.0, 0.0, 0.0, 0.0;
  freq << 0.2, 0.0, 0.0, 0.0, 0., 0.0;

  while (ros::ok())
  {

    //sine reference
    q_des = q_des0.array() + amp.array()*(2*M_PI*freq*loop_time).array().sin();

    loop_time += (double)1/loop_frequency;
    //send_des_jstate_sim(q_des);
    send_des_jstate(q_des);
    loop_rate.sleep();
  }

  return 0;
}

File: fdr_exercises/custom_joint_publisher.py
#!/usr/bin/env python
import rospy as ros
import numpy as np
from std_msgs.msg import Float64MultiArray


class JointStatePublisher():

    def __init__(self):
        self.q_des =np.zeros(6)
        self.qd_des = np.zeros(6)
        self.tau_ffwd = np.zeros(6)
        self.filter_1 = np.zeros(6)
        self.filter_2 = np.zeros(6)

    def send_des_jstate(self):
        msg = Float64MultiArray()
        if (self.gripper_sim):
            msg.data = np.append(self.q_des, np.array([0.0, 0.0 ,0.0]))
        else:
            msg.data = self.q_des
        self.pub_des_jstate.publish(msg)

def talker(p):
    ros.init_node('custom_joint_pub_node', anonymous=True)
    p.pub_des_jstate = ros.Publisher("/ur5/joint_group_pos_controller/command", Float64MultiArray, queue_size=1)
    loop_frequency = 1000.
    loop_rate = ros.Rate(loop_frequency)  # 1000hz

    # init variables
    time = 0
    q_des0 = np.array([ -0.32,-0.78, -2.56,-1.63, -1.57, 3.49])

    # 1 fixed setpoint
    p.q_des = q_des0

    # check if gripper is actuated
    p.gripper_sim = ros.get_param("/gripper_sim")

    amp = np.array([0.3, 0.0, 0.0, 0.0, 0.0, 0.0])  # amplitude
    freq = np.array([0.2, 0.0, 0.0, 0.0, 0., 0.0]) # frequency

    while not ros.is_shutdown():
        if p.gripper_sim == True:
            print("this publisher cannot handle the gripper joints")
            break
        # 2 - generate step reference
        # if time < 4.:
        #     p.q_des = q_des0
        # else:
            #p.q_des = q_des0 + np.array([0., 0.4, 0., 0., 0., 0])


        p.qd_des = np.zeros(6)
        p.tau_ffwd = np.zeros(6)

        p.send_des_jstate()
        print(p.q_des)
        time = np.round(time + np.array([1/loop_frequency]), 3)
        loop_rate.sleep()

if __name__ == '__main__':
    myPub = JointStatePublisher()
    try:
        talker(myPub)
    except ros.ROSInterruptException:
        pass
